txpipe/noise_maps.py

```diff
@@ -541,3 +541,216 @@ def finalize_mappers(self, pixel_scheme, mappers):
+        ('mask', MapsFile),
+        ('lens_maps', MapsFile),
+    ]
+    outputs = [
+        ('source_noise_maps', LensingNoiseMaps),
+        ('lens_noise_maps', ClusteringNoiseMaps),
+    ]
+
+    config_options = {
+        'chunk_rows': 2000000,
+        'lensing_realizations': 30,
+        'clustering_realizations': 1,
+    }
+
+    def run(self):
+        from jax import numpy as jnp
+        from jax.ops import index
+        from jax import random, jit, device_get, device_put
+        from .utils import choose_pixelization
+        # get the number of bins.
+        nbin_source, nbin_lens, ngal_maps, mask, map_info = self.read_inputs()
+        pixel_scheme = choose_pixelization(**map_info)
+        lensing_realizations = self.config['lensing_realizations']
+        clustering_realizations = self.config['clustering_realizations']
+
+        # The columns we will need
+        shear_cols = ['ra', 'dec', 'weight', 'mcal_g1', 'mcal_g2']
+
+        # Make the iterators
+        chunk_rows = self.config['chunk_rows']
+
+        it = self.combined_iterators(chunk_rows,
+                'shear_catalog', 'shear', shear_cols,
+                'shear_tomography_catalog','tomography', ['source_bin'],
+                'lens_tomography_catalog','tomography', ['lens_bin'],
+            )
+
+        # Get a mapping from healpix indices to masked pixel indices
+        # This reduces memory usage.  We could use a healsparse array
+        # here, but I'm not sure how to do that best with our
+        # many realizations.  Possiby a recarray?
+
+        index_map = np.zeros(pixel_scheme.npix, dtype=jnp.int32) - 1
+        counter = 0
+        for i in range(pixel_scheme.npix):
+            if mask[i] > 0:
+                index_map[i] = counter
+                counter += 1
+        # Number of unmasked pixels
+        npix = counter
+
+        if self.rank == 0:
+            nmaps = nbin_source * (2 * lensing_realizations + 1) + nbin_lens * clustering_realizations * 2
+            nGB = (npix * nmaps * 8) / 1000.**3
+            print(f"Allocating maps of size {nGB:.2f} GB")
+        # lensing g1, g2
+        G1 = jnp.zeros((npix, nbin_source, lensing_realizations))
+        G2 = jnp.zeros((npix, nbin_source, lensing_realizations))
+        # lensing weight
+        GW = jnp.zeros((npix, nbin_source))
+        # clustering map - n_gal to start with
+        ngal_split = jnp.zeros((npix, nbin_lens, clustering_realizations, 2), dtype=np.int32)
+        # TODO: Clustering weights go here
+
+        # Initialize PRNG key for Jax
+        key = random.PRNGKey(np.random.randint(2**32))
+
+        # Loop through the data
+        for (s, e, data) in it:
+            print(f"Rank {self.rank} processing rows {s} - {e}")
+            # Send data to GPU
+            data = device_put(data)
+            source_bin = data['source_bin']
+            lens_bin = data['lens_bin']
+            ra = data['ra']
```

Comment:
That is correct. I think I'll just manually select what things to send to GPU instead of sending the whole of data since ra and dec are not used after that step.